Solution 작성, 수정, 삭제 기능

jd99iam opened the pull request:
## Description
Solution 작성, 수정, 삭제 기능을 개발했습니다.

## Changes
- [x] Solution 작성 기능 개발 
- [x] Solution 수정 기능 개발 
- [x] Solution 삭제 기능 개발 

## To Reviewers
예외처리를 어떤 식으로 하는게 좋을까요?

jd99iam committed:
[#1]feat: Solution Entity 추가

jd99iam committed:
[#1]feat: Solution Entity 수정

jd99iam committed:
[#1]feat: Problem Entity 추가

jd99iam committed:
[#1]feat: Member Entity 추가

jd99iam committed:
[#1]feat: Like Entity 추가

jd99iam committed:
[#1]feat: Solution Entity 수정

jd99iam committed:
[#1]feat: Algorithm Entity 추가

jd99iam committed:
[#1]feat: DataStructure Entity 추가

jd99iam committed:
[#1]feat: SolutionController 추가

jd99iam committed:
[#1]feat: SolutionService 추가

jd99iam committed:
[#1]feat: SolutionRepository 추가

jd99iam committed:
[#1]feat: Solution Entity에 Level 추가

jd99iam committed:
[#1]feat: enum Level 추가

jd99iam committed:
[#1]feat: SolutionRequest 추가

jd99iam committed:
[#1]feat: ErrorResponse 추가

jd99iam committed:
[#1]fix: 엔티티 매핑 수정

jd99iam committed:
[#1]feat: Solution Entity에 toEntity 메소드 추가

jd99iam committed:
[#1]feat: SolutionService 정상로직 추가

jd99iam committed:
[#1]feat: BackendApplication 보안 설정 끄기

jd99iam committed:
[#1]fix: Solution Entity에 cascade 속성 추가

jd99iam committed:
[#1]fix: Solution Entity 필드에 cascade 추가, toEntity 수정

jd99iam committed:
[#1]feat: SolutionRequest에 Valid message 추가

jd99iam committed:
[#1]fix: SolutionController, ExControllerAdvice 수정

jd99iam committed:
[#1]test: SolutionRepository save 테스트 추가

jd99iam committed:
[#1]test: SolutionService 저장 테스트 추가

jd99iam committed:
[#1]feat: SolutionController 수정 폼 컨트롤러 추가

jd99iam committed:
[#1]feat: SolutionUpdateForm 추가

jd99iam committed:
[#1]feat: ExControllerAdvice noSuchElementEx 추가

jd99iam committed:
[#1]fix: ErrorResponse에 Getter 추가

jd99iam committed:
[#1]feat: SolutionPatchRequest 추가

jd99iam committed:
[#1]refactor: SolutionController 코드 수정

jd99iam committed:
[#1]fix: SolutionService 코드 수정

jd99iam committed:
[#1]feat: SolutionController delete 추가

jd99iam committed:
[#1]fix: SolutionService delte 수정

jd99iam committed:
[#1]test: SolutionRepositoryTest 수정 테스트 추가

jd99iam committed:
[#1]test: SolutionRepository 삭제 테스트 추가

jd99iam committed:
[#1]refactor: SolutionController의 예외처리 로직 SolutionService로 이동

jd99iam committed:
[#1]feat: Errorcode와 SolutionNotFoundException 추가

jd99iam committed:
[#1]fix: enum 수정

jd99iam committed:
[#1]refactor: Solution에 update 메소드 추가, toEntity 삭제

jd99iam committed:
[#1]feat: SolutionRequest에 toEntity 추가

jd99iam committed:
[#1]feat: Solution에 update 메소드 추가

jd99iam committed:
[#1]fix: enum으로 받는 dto에 Valid 애노테이션 제거

jd99iam committed:
[#1]feat: Member에 update 메소드 추가

jd99iam committed:
[#1]feat: Problem에 update 메소드 추가

jd99iam committed:
[#1]refactor: 에러 타입 변경, 핸들러 내용 수정

jd99iam committed:
[#1]feat: 문제 링크 검증 및 에러 추가

jd99iam committed:
[#1]fix: 저지명 enum으로 변경

jd99iam committed:
[#1]fix: Judges enum type으로 수정

jd99iam committed:
[#1]feat: 스크랩 해서 Solution 생성하는 기능 추가

jd99iam committed:
[#1]fix: Scrap해서 Solution 생성하는 기능 수정

jd99iam committed:
[#1]refactor: SolutionService findById에서 예외처리하도록 수정

jd99iam committed:
[#1]refactor: SolutionController 예외 throw 삭제

jd99iam committed:
test: SolutionRepositoryTest 테스트 코드 전체 수정

1. SolutionRepositoryTest에서 @DataJpaTest로 전환
2. 솔루션_수정_테스트 코드 새로운 엔티티를 생성하지  않고 findById로 찾은 엔티티에 update필드명() 메소드로 변경하도록 수정
3. 솔루션_삭제_테스트에서 체크하려는 에러 타입 변경

Related to : #1

jd99iam committed:
test: SolutionServiceTest 저장_테스트 수정

1. Mokito를 사용하도록 수정
2. given 사용하도록 수정

Related to : #1

jd99iam committed:
test: SolutionFixture 추가

1. SolutionFixture 추가

Related to : #1

jd99iam committed:
fix: Solution에서 likes 필드 삭제

1. Solution에서 likes 필드 삭제

Related to : #1

jd99iam committed:
feat: SolutionServiceTest 수정_테스트 추가

1. SolutionServiceTest 수정_테스트 추가

Related to : #1

jd99iam committed:
fix: Like를 Likes로 rename

1. Like를 Likes로 rename

Related to : #1

jd99iam committed:
fix: Fixture 대신 Builder로 테스트하도록 변경

1. Fixture 대신 Builder로 테스트하도록 변경

Related to : #1

jd99iam committed:
fix: Solution에 updateLike 메소드 제거

1. Fixture 대신 Builder로 테스트하도록 변경

Related to : #1

jd99iam committed:
fix: database 수정

1. application.properties에 prograngers2 database 사용하도록 수정

Related to : #1

jd99iam committed:
fix: SolutionFixture 수정

1. SolutionFixture에서 객체 필드는 빌더를 사용하도록 변경

Related to : #1

jd99iam committed:
test: SolutionServiceTest 수정_테스트 수정

1. SolutionServiceTest에서 repository를 테스트하도록 잘못 작성한 코드 수정

Related to : #1

jd99iam committed:
fix: @GeneratedValue 전략 IDENTITY로 수정

1. @GeneratedValue 전략을 IDENTITY로 수정함

Related to : #1

jd99iam committed:
test: SolutionServiceTest 삭제 테스트 추가

1. verify를 활용해 삭제 테스트를 추가함

Related to : #1

jd99iam committed:
style: SolutionRepositoryTest, SolutionService 스타일 수정

1. SolutionService 들여쓰기 수정
2. SolutionRepositoryTest import문 수정

Related to : #1

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/BackendApplication.java:
```diff
@@ -2,8 +2,9 @@
+import org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration;
 
-@SpringBootApplication
+@SpringBootApplication(exclude = { SecurityAutoConfiguration.class })
```
exclude는 security 인증 때문에 그런걸까요 ?

jd99iam replied:
> exclude는 security 인증 때문에 그런걸까요 ?

네 맞습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/ExControllerAdvice.java:
```diff
@@ -0,0 +1,53 @@
+
+    // Valid를 통과하지 못할 경우 ErrorResponse dto로 해당하는 에러 반환
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<List<ErrorResponse>> validEx(MethodArgumentNotValidException exception){
```
메서드 명을 더 구체적으로 변경하면 좋을 것 같아요
그리고 Exception을 Ex로 줄여쓰지 말기 ~~

jd99iam replied:
> 메서드 명을 더 구체적으로 변경하면 좋을 것 같아요 그리고 Exception을 Ex로 줄여쓰지 말기 ~~

notValidException으로 변경했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/ExControllerAdvice.java:
```diff
@@ -0,0 +1,53 @@
+
+     // 요청한 PathVariable에 대한 데이터가 db에 없는 경우 에러메세지와 함께 NoSuchElementException 던짐
+    @ExceptionHandler
+    public ResponseEntity<ErrorResponse> noSuchElementEx(NotFoundException exception){
```
이것도 메서드명을 not found와 관련된 메서드 임을 알 수 있도록 변경해주세요 ~~

jd99iam replied:
> 이것도 메서드명을 not found와 관련된 메서드 임을 알 수 있도록 변경해주세요 ~~

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/ExControllerAdvice.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+     // 요청한 PathVariable에 대한 데이터가 db에 없는 경우 에러메세지와 함께 NoSuchElementException 던짐
+    @ExceptionHandler
```
위에선 @ExcetpionHandler에 값이 들어갔으니 전부 통일해서 여기도 넣는게 좋을 것 같아요
캐스팅만 믿는건 위험하기도 하고 직관적이지 못한 것 같아요 !

jd99iam replied:
> 위에선 @ExcetpionHandler에 값이 들어갔으니 전부 통일해서 여기도 넣는게 좋을 것 같아요 캐스팅만 믿는건 위험하기도 하고 직관적이지 못한 것 같아요 !

반영했습니다!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -0,0 +1,94 @@
+
+    // solution 쓰기
+    @PostMapping("/new-form")
+    public ResponseEntity<?> newForm(@RequestBody @Valid SolutionRequest solutionRequest) throws URISyntaxException {
```
throwException들은 없애기로 했던 것 같은데 남아있는 이유가 있나요 혹시?

jd99iam replied:
URISyntaxException은 리다이렉트 할 때 URI를 생성하려면 필요해서 넣었습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -0,0 +1,94 @@
+
+        // 리포지토리 활용해 저장
+        Solution solution = solutionRequest.toEntity();
+        Solution saved = solutionService.save(solution);
```
inline으로 변수를 합치는게 낫지 않을까요 ?

jd99iam replied:
> inline으로 변수를 합치는게 낫지 않을까요 ?

반영했습니다

rlfrkdms1 replied:
무작정 반영하겠다가 아니라 지담님의 의견이 궁금해요
저는 그냥 제시해보는거에요 ! 다음엔 어떤 의견을 가지고 있는지도 적어주세요 !

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -0,0 +1,94 @@
+        Long solutionId = saved.getId();
+
+        // 성공할 시 solutiuonId에 해당하는 URI로 리다이렉트, 상태코드 302
+        URI redirectUri = new URI("http://localhost:8080/solutions/"+solutionId);
```
여기도 합칠 수 있을 것 같아요 !
그리고 redirect uri 같은 경우는 상수를 이용해서 관리하는게 어떨까요 ?

jd99iam replied:
> 여기도 합칠 수 있을 것 같아요 ! 그리고 redirect uri 같은 경우는 상수를 이용해서 관리하는게 어떨까요 ?

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -0,0 +1,94 @@
+    // 수정 폼 반환
+    @GetMapping("/{solutionId}/update-form")
+    public ResponseEntity<?> updateForm(@PathVariable Long solutionId){
+        Solution target = solutionService.findById(solutionId);
```
contoller에는 최대한 엔티티를 드러내지 않았으면 좋겠어요 !
차라리 service에서 수정 폼 자체로 변경해서 반환하는건 어떨까요?

jd99iam replied:
> contoller에는 최대한 엔티티를 드러내지 않았으면 좋겠어요 ! 차라리 service에서 수정 폼 자체로 변경해서 반환하는건 어떨까요?

컨트롤러에서 엔티티를 드러내면 안좋은 이유가 뭔가요??

rlfrkdms1 replied:
제가 슬랙에 관련 자료 올려뒀으니 한번 읽어주시겠어요 ? 
사실 정답은 없는 문제라 수요일에 같이 얘기해보고 통일하면 좋을 것 같아요 !

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/ErrorResponse.java:
```diff
@@ -0,0 +1,15 @@
+   // String description;
+
+    ErrorCode errocode;
+    String descriptions;
```
List로 관리한다면 descriptions가 아니라 description이 낫지 않을까요 ?

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/ScarpSolutionRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+    @NotBlank
+    private String description;
+    private Levels level;
```
Enum class 명이 Levels인 이유가 있나요 ? Level이어도 괜찮을 것 같아요!

jd99iam replied:
> Enum class 명이 Levels인 이유가 있나요 ? Level이어도 괜찮을 것 같아요!

Algorithm 엔티티가 있고 엔티티에서 필드로 사용하는 Algorithms enum이 있고 DataStructure도 상황이 같아서 enum 뒤에 구분을 위해 enum을 전부 복수형 이름으로 지었습니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionPatchRequest.java:
```diff
@@ -0,0 +1,37 @@
+    @NotBlank(message = "문제 제목을 입력해주세요")
+    private String title;
+
+   //  @NotBlank(message = "알고리즘 이름을 입력해 주세요")
```
이렇게 제가 볼 필요 없는 주석들은 지우고 push 해주세요 !

jd99iam replied:
> 이렇게 제가 볼 필요 없는 주석들은 지우고 push 해주세요 !

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionPatchRequest.java:
```diff
@@ -0,0 +1,37 @@
+    private Algorithms algorithmName;
+
+    // @NotBlank(message = "자료구조 이름을 입력해 주세요")
+    private DataStructures dataStructureName;
```
만약 null 값이 허용 된다면 @Nullable로 null 값을 허용한다고 직관적으로 표현하는 건 어떨까요?
그리고 변수명에 그냥 name을 사용하지 않고 algorithm으로 쓰는 건 어떨까요 ?

jd99iam replied:
> 만약 null 값이 허용 된다면 @nullable로 null 값을 허용한다고 직관적으로 표현하는 건 어떨까요? 그리고 변수명에 그냥 name을 사용하지 않고 algorithm으로 쓰는 건 어떨까요 ?

필드 이름은 정의한 API에 맞춘 것입니다. 그리고 null값을 허용하지 않고 enum type에서 @JsonCreater에서 자체적으로 검증이 되기 때문에 @NotBlank를 지웠습니다.

rlfrkdms1 replied:
algorithm으로 변경하는거에 대한 의견이 궁금합니다
무조건 api 명세서를 따라야할 필요도 없고 api 명세는 계속 변동 될 수 있는거니까요
그냥 어떤걸 따랐다 ~가 그래서 그랬다 이런것 보단 지담님의 의견이 궁금해요

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionRequest.java:
```diff
@@ -0,0 +1,73 @@
+        Judges judge = checkLink(problemLink);
+
+        return Solution.builder()
+                .problem(new Problem(null, problemTitle, problemLink, judge)) // 파싱해서 ojname 알아내야함
```
여기선 builder를 안쓰고 생성자를 사용한 이유가 있나요?

jd99iam replied:
> 여기선 builder를 안쓰고 생성자를 사용한 이유가 있나요?

코드가 너무 길어질 것 같아서 생성자를 썼는데 builder를 써야 한다면 수정하겠습니다

rlfrkdms1 replied:
그렇군요 저는 뭔가 builder나 생성자 둘 중에 하나로 통일하고 싶어서 말씀드렸어요 수요일에 얘기해보고 통일해봐요

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionRequest.java:
```diff
@@ -0,0 +1,73 @@
+                .build();
+    }
+
+    private Judges checkLink(String problemLink) {
```
굳이 메서드를 하나 더 뺀 이유가 있나요 ? 체크하고 있음을 알리는 걸까요 ?

jd99iam replied:
> 굳이 메서드를 하나 더 뺀 이유가 있나요 ? 체크하고 있음을 알리는 걸까요 ?

네 맞습니다 알아보기 쉽게 하려고 이렇게 작성했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionUpdateForm.java:
```diff
@@ -0,0 +1,34 @@
+    private String link;
+    private Algorithms algorithmName;
+    private DataStructures dataStructureName;
+
```
띄어쓰기가 왜 있을까용

jd99iam replied:
> 띄어쓰기가 왜 있을까용

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/SolutionUpdateForm.java:
```diff
@@ -0,0 +1,34 @@
+import com.prograngers.backend.entity.Algorithms;
+import com.prograngers.backend.entity.DataStructures;
+import com.prograngers.backend.entity.Solution;
+import lombok.*;
```
와일드 카드가 사용되었네요 !

jd99iam replied:
> 와일드 카드가 사용되었네요 !

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/Algorithms.java:
```diff
@@ -0,0 +1,39 @@
+    DFS("DFS"),
+    DIJKSTRA("다익스트라");
+
+    private final String krName;
```
kr이 뭔가요 ? 한글 이름이라는 걸까요?
음,, krName 말고 다른 변수명으로 생각해보면 좋을 것 같아요

jd99iam replied:
> kr이 뭔가요 ? 한글 이름이라는 걸까요? 음,, krName 말고 다른 변수명으로 생각해보면 좋을 것 같아요

해당 enum값의 한글이름 이라는 뜻 맞습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/DataStructure.java:
```diff
@@ -0,0 +1,22 @@
+package com.prograngers.backend.entity;
+
+import jakarta.persistence.*;
```
와일드 카드가 사용되었네요 ~

jd99iam replied:
> 와일드 카드가 사용되었네요 ~

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/DataStructure.java:
```diff
@@ -0,0 +1,22 @@
+    private Long id;
+
+    @Enumerated(EnumType.STRING)
+    private DataStructures name;
```
엔티티 이름이 name으로 되어있네요 이러면 테이블에 저장될 때 name으로 저장될 것 같은데 어떤가요 ?
만약 name으로 저장된다면 안좋은 방법인 것 같아요

jd99iam replied:
> 엔티티 이름이 name으로 되어있네요 이러면 테이블에 저장될 때 name으로 저장될 것 같은데 어떤가요 ? 만약 name으로 저장된다면 안좋은 방법인 것 같아요

DataStructure 테이블의 name으로 저장되면 자료구조의 이름을 뜻하니까 맞다고 생각했는데dataStructureName 과 같은 이름이 낫다는 말씀이신가요?

rlfrkdms1 replied:
또 그렇게 생각하니까 그렇네요 헤헤

rlfrkdms1 replied:
음 고민이 조금 필요할 것 같아요 수요일에 얘기해봐요 !!

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/DataStructures.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public enum DataStructures {
```
DataStructureConstant 와 같은 방식으로 변수명 작명은 어떤가요 ?
이것도 괜찮은 방법 같긴한데 조금 고려해볼 순 있을 것 같아요

jd99iam replied:
> DataStructureConstant 와 같은 방식으로 변수명 작명은 어떤가요 ? 이것도 괜찮은 방법 같긴한데 조금 고려해볼 순 있을 것 같아요

좋은 생각인 것 같습니다. enum 타입들을 constant 패키지로 묶고 이름을 말씀하신대로 모두 변경했습니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/Judges.java:
```diff
@@ -0,0 +1,28 @@
+    백준("acmicpc"),
+    프로그래머스("programmers");
+
+    private final String link;
```
사실상 위의 acmicpc와 programmers는 링크가 아니어서 value나 다른 이름으로 작명하는게 좋을 것 같아요 !

jd99iam replied:
> 사실상 위의 acmicpc와 programmers는 링크가 아니어서 value나 다른 이름으로 작명하는게 좋을 것 같아요 !

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/Solution.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public void updateAlgorithm(Algorithms algorithm){
+        this.algorithm = new Algorithm(null,algorithm);
```
왜 생성자를 사용한건가요 ?

jd99iam replied:
> 왜 생성자를 사용한건가요 ?

알고리즘, 자료구조 업데이트 메소드 빌더로 수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/exception/ErrorCode.java:
```diff
@@ -0,0 +1,13 @@
+    ALGORITHM_NOT_EXISTS,
+    DATASTRUCTURE_NOT_EXISTS,
+    LEVEL_NOT_EXISTS,
+
```
띄어쓰기 고쳐주세요!

jd99iam replied:
> 띄어쓰기 고쳐주세요!

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/exception/ErrorCode.java:
```diff
@@ -0,0 +1,13 @@
+public enum ErrorCode {
+
+    INVALID_SOLUTION_BODY,
+    SOLUTION_NOT_EXISTS,
```
NOT_EXIST는 not found를 말하는 걸까요 ? 다 통일하면 좋을 것 같아요

jd99iam replied:
> NOT_EXIST는 not found를 말하는 걸까요 ? 다 통일하면 좋을 것 같아요

수정했습니다.
NOT_FOUND : NotFoundException을 상속받는 예외의 에코드 
NOT_EXISTS : EnumTypeException을 상속받는 예외의 에러코드

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/exception/ErrorCode.java:
```diff
@@ -0,0 +1,13 @@
+    DATASTRUCTURE_NOT_EXISTS,
+    LEVEL_NOT_EXISTS,
+
+    PROBLEM_LINK_NOT_FOUND;
```
이건 NOT_FOUND 네요 ? 혹시 위의 NOT_EXISTS와 다른 점이 있는걸까요 ?

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/exception/enumtype/EnumTypeException.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+public class EnumTypeException extends RuntimeException{
+    ErrorCode errorCode;
```
final 로 변경해주면 좋을 것 같아요 그리고 CustomException을 만들어서 format 관리를 해주세요 !

jd99iam replied:
> final 로 변경해주면 좋을 것 같아요 그리고 CustomException을 만들어서 format 관리를 해주세요 !

final로 변경했습니다. CustomException을 만들어서 format 관리를 해주세요 는 이해가 잘 가지 않습니다..

rlfrkdms1 replied:
그럼 이건 수요일에 같이 수정할게요 ~~

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/exception/enumtype/EnumTypeException.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+public class EnumTypeException extends RuntimeException{
```
생각해보면 EnumTypeException이라는 이름은 직관적이지 못한 것 같아요
다른 식의 작명도 고려해보면 좋을 것 같습니다 어찌보면 not found와 유사하고 어찌보면 입력 형식의 문제와 유사한 것 같아서요 !

jd99iam replied:
> 생각해보면 EnumTypeException이라는 이름은 직관적이지 못한 것 같아요 다른 식의 작명도 고려해보면 좋을 것 같습니다 어찌보면 not found와 유사하고 어찌보면 입력 형식의 문제와 유사한 것 같아서요 !



> 생각해보면 EnumTypeException이라는 이름은 직관적이지 못한 것 같아요 다른 식의 작명도 고려해보면 좋을 것 같습니다 어찌보면 not found와 유사하고 어찌보면 입력 형식의 문제와 유사한 것 같아서요 !

알고리즘, 자료구조, 레벨과 같은 enum에서 dto로 들어온 값을 변환하다가 생기는 예외를  모두 EnumTypeException을 상속하게 해서 괜찮다고 생각합니다. 입력 형식에 문제가 있는 것 중 enum과 관련된 것들입니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+@RequiredArgsConstructor
+@Service
+@Slf4j
+public class SolutionService {
```
서비스단에 @Transactional 이 없는 것 같아요 ! 트랜잭션 관리는 항상 해주세요

rlfrkdms1 replied:
클래스 단에는 readonly = true로 해주시고 메서드 단에서 false를 관리해주세요 !

jd99iam replied:
> readonly = true

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+        return saved;
+    }
+
+    public Solution update(Long solutionId, SolutionPatchRequest request) throws SolutionNotFoundException {
```
throws를 삭제하기로 했던 것 같은데 있는 이유가 있나요 ?

jd99iam replied:
> throws를 삭제하기로 했던 것 같은데 있는 이유가 있나요 ?

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/SolutionController.java:
```diff
@@ -0,0 +1,94 @@
+                                    @RequestBody @Valid SolutionPatchRequest solutionPatchRequest) throws URISyntaxException {
+
+        // solutionService로 update한다
+        Solution updated = solutionService.update(solutionId,  solutionPatchRequest);
```
updated는 사용되지 않는걸로 보이는데 왜 반환받는지 모르겠어요 이유가 있나요?

jd99iam replied:
> updated는 사용되지 않는걸로 보이는데 왜 반환받는지 모르겠어요 이유가 있나요?

updated.getId()로 리다이렉 URI를 얻습니다. 해당 값은 PathVariable로 얻어도 무관하긴 합니다. 나중에 반환값이 필요해질 수도 있다고 생각해서 반환하도록 만들었습니다.

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+        return updated;
+    }
+
+//    public void delete(Solution solution){
```
상관없는 주석은 삭제해주세요 !

jd99iam replied:
> 상관없는 주석은 삭제해주세요 !

반영했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+            solutionRepository.delete(target);
+    }
+
+    public List<Solution> index(Member member){
```
index 보다는 조금더 직관적인 이름으로 작명하는게 어떨까요 ?
이것만 보고는 어떤 메서드인지 잘 모르겠어요!

jd99iam replied:
> index 보다는 조금더 직관적인 이름으로 작명하는게 어떨까요 ? 이것만 보고는 어떤 메서드인지 잘 모르겠어요!

해당 메소드는 사용하지 않아서 삭제했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+
+    public List<Solution> index(Member member){
+
+        List<Solution> solutionList = solutionRepository.findAllByMember(member);
```
만약 member가 없을 경우에 예외는 없는걸까요 ?

jd99iam replied:
> 만약 member가 없을 경우에 예외는 없는걸까요 ?

해당 메소드는 사용하지 않아서 삭제했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -0,0 +1,85 @@
+                .code(scrap.getCode())
+                .scraps(0)
+                .scrapId(scrap)
+                .algorithm(new Algorithm(null, scrap.getAlgorithm().getName()))
```
여기도 생성자네요
그리고 생각해보면 algorithm과 data structure가 하나씩 만 들어가면 따로 테이블을 만들어서 관리하는 이유가 없는 것 같아요
이부분 변경할까요 ?

jd99iam replied:
> 여기도 생성자네요 그리고 생각해보면 algorithm과 data structure가 하나씩 만 들어가면 따로 테이블을 만들어서 관리하는 이유가 없는 것 같아요 이부분 변경할까요 ?

저도 왜 따로 테이블을 만들어야 하는지 잘 이해가 가지 않았습니다. 변경해도 좋을 것 같습니다

rlfrkdms1 replied:
이부분 그럼 수요일에 같이 얘기나눠봐요

rlfrkdms1 commented on backend/src/main/resources/application.properties:
```diff
@@ -1 +1,8 @@
+spring.datasource.username=root
+spring.datasource.password=0000
+
+# jpa?? ???? ??? ????? ?? ?? update
```
삭제해주세요 ~

jd99iam replied:
> 삭제해주세요 ~

반영했습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/repository/SolutionRepositoryTest.java:
```diff
@@ -0,0 +1,99 @@
+    @Test
+    void 정상_입력_저장_테스트(){
+
+//        // given
```
주석 ㅜㅜ

jd99iam replied:
> 주석 ㅜㅜ

삭제했습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/repository/SolutionRepositoryTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Test
+    @Transactional
+    void 솔루션_삭제_테스트(){
```
지금 메서드 명에는 삭제 테스트라고 되어있는데 밑에서 예외를 던지고 있어요
예외 테스트는 따로 작성하는게 좋을 것 같아요 !

jd99iam replied:
> 지금 메서드 명에는 삭제 테스트라고 되어있는데 밑에서 예외를 던지고 있어요 예외 테스트는 따로 작성하는게 좋을 것 같아요 !

findById로 조회했을 때 삭제되어서 null인 경우 예외를 던지는 식으로 테스트 하면 된다고 생각해서 이렇게 작성했습니다.

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/service/SolutionServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+@Slf4j
+class SolutionServiceTest {
+    // Mokito , SpringBootTest 빼고
+    @Mock
```
line 정리를 항상 해주시는게 좋을 것 같아요 ~ 단축키를 찾아보는게 어떨까요 !

jd99iam replied:
> line 정리를 항상 해주시는게 좋을 것 같아요 ~ 단축키를 찾아보는게 어떨까요 !

모든 파일에 대해 라인 정리 반영했습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/service/SolutionServiceTest.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Test
+    void 수정_테스트(){
```
테스트명 작성에 대해서 얘기를 나눠봐야할 것 같아요 조금 더 자세하면 좋을 것 같아요
수정의 어떤 테스트인지 어떤 상황에 대한 테스트인지 확인하기 어려워요 ㅜㅜ

jd99iam replied:
> 테스트명 작성에 대해서 얘기를 나눠봐야할 것 같아요 조금 더 자세하면 좋을 것 같아요 수정의 어떤 테스트인지 어떤 상황에 대한 테스트인지 확인하기 어려워요 ㅜㅜ

토의 후 테스트 코드 전체적으로 수정하도록 하겠습니다

rlfrkdms1 reviewed (commented):
전체적으로 컨벤션이 잘 지켜지지 않은 것 같아요 ! 
코드 작성하거나 커밋을 할 때 한번씩 확인해보는게 어떨까요 ?
와일드 카드 같은 경우엔 인텔리제이 설정으로 관리하시면 될 것 같아요 ! :)
리뷰 드리는건 제 의견을 따르라는게 아닌 그저 제 의견입니다 ! 
한번씩 읽어보시고 지담님은 왜 그렇게 하셨는지와 의견을 달아주시면 좋을 것 같아요 :)

rlfrkdms1 reviewed (commented):
전체적으로 컨벤션이 잘 지켜지지 않은 것 같아요 ! 
코드 작성하거나 커밋을 할 때 한번씩 확인해보는게 어떨까요 ?
와일드 카드 같은 경우엔 인텔리제이 설정으로 관리하시면 될 것 같아요 ! :)
리뷰 드리는건 제 의견을 따르라는게 아닌 그저 제 의견입니다 ! 
한번씩 읽어보시고 지담님은 왜 그렇게 하셨는지와 의견을 달아주시면 좋을 것 같아요 :)

jd99iam committed:
refactor: ExControllerAdvice에서 notValidException으로 메소드 명 변경

1. notValidException으로 더 구체적인 메소드 명으로 변경했습니다.

Related to : #1

jd99iam committed:
refactor: ExControllerAdvice에서 메소드 명 변경

1. 메소드 명으로 예외 타입을 알 수 있도록 변경했습니다.
2. ex로 줄여쓰지 않고 exception으로 변경했습니다

Related to : #1

jd99iam committed:
refactor: ExControllerAdvice에서 애노테이션에 클래스명 추가

1. 직관적으로 알 수 있도록 하기 위해 @ExceptionHandler 애노테이션에 예외 클래스 타입을 추가했습니다

Related to : #1

jd99iam committed:
refactor: SolutionController 리팩토링

1. 인라인으로 변경
2. REDIRECT_PATH 상수로 관리

Related to : #1

jd99iam committed:
refactor: SolutionPatchRequest 주석 삭제

1. 불필요한 주석 삭제

Related to : #1

jd99iam committed:
refactor: 띄어쓰기 삭제

1. SolutionUpdateForm에 불필요한 개행 삭제

Related to : #1

jd99iam committed:
refactor: 와일드카드 수정

1. 와일드카드 사용하지 않도록 수정

Related to : #1

jd99iam committed:
refactor: DataStructure

1. 와일드카드 사용하지 않도록 수정

Related to : #1

jd99iam committed:
refactor: enum 패키지 이동 및 이름 수정

1. enum 클래스들 패키지를 constant 패키지로 이동함
2. 이름을 ~Constant 형태로 수정함

Related to : #1

jd99iam committed:
refactor: JudgeConstant 의 link를 value로 수정

1. link가 아닌 링크에 포함된 내용이므로 value로 수정

Related to : #1

jd99iam committed:
refactor: Solution의 update 메소드에서 builder를 사용하도록 변경

1. 알고리즘, 자로구조 업데이트 시 생성자가 아닌 빌더  사용하도록 변경

Related to : #1

jd99iam committed:
refactor: SOLUTION_NOT_FOUND로 rename

1. SOLUTION_NOT_EXISTS -> SOLUTION_NOT_FOUND로 rename

Related to : #1

jd99iam committed:
refactor: 예외에서 ErrorCode 필드 final로 변경

1. EnumTypeException에서 errorCode 필드 final로 변경
2. NotFoundException에서 errorCode 필드 final로 변경

Related to : #1

jd99iam committed:
refactor: SolutionService에 Transactional 추가

1. Transactional 추가
2. 사용하지 않는 index 메소드 삭제

Related to : #1

jd99iam committed:
refactor: SolutionService에 throws 삭제

1. SolutionService update에 throws 삭제

Related to : #1

jd99iam committed:
refactor: application.properties에 주석 삭제

1. application.properties에 필요없는 주석 삭제

Related to : #1

jd99iam committed:
refactor: 라인 정리

1. 모든 파일에 대해 라인 정리

Related to : #1

rlfrkdms1 commented:
고생 많으셨습니다 !!
처음 해보는 과정이라 많이 생소하기도 하고 번거로운 작업도 많았을텐데, 잘 따라와주셔서 감사합니다 
제가 남긴 comment들 마저 확인하신 후에 MERGE 하시고 브랜치 삭제 해주시면 되겠습니다 !
수요일에 의논해볼 것들은 수요일에 만나서 얘기나눠 봅시다
남은 기능들도 화이팅이에요 !! 😀😀😀

jd99iam commented:
> 고생 많으셨습니다 !! 처음 해보는 과정이라 많이 생소하기도 하고 번거로운 작업도 많았을텐데, 잘 따라와주셔서 감사합니다 제가 남긴 comment들 마저 확인하신 후에 MERGE 하시고 브랜치 삭제 해주시면 되겠습니다 ! 수요일에 의논해볼 것들은 수요일에 만나서 얘기나눠 봅시다 남은 기능들도 화이팅이에요 !! 😀😀😀

자세한 comment 감사합니다. merge 하겠습니다
